feat: fast url entry

terranblake opened the pull request:
This attempts to resolve the feature request described in the [Fast Url Entry](https://github.com/liyasthomas/postwoman/issues/62) issue. This was accomplished by removing the UI references to path, moving fields to allow for more URL space and by updating the 1 failing test to exclude the path field

* [x]  Remove reference to path in ui
* [x]  Inject query parameters on URL field change
* [x]  Move `Permalink` button to bottom row
* [x]  Move `Show Code` button to bottom row

terranblake commented:
@liyasthomas let me know if there's anything I can help with on this. I ran the provided tests locally, but am not familiar enough with this stack/repo, so I wanted to check in an make sure 👍

liyasthomas commented:
@terranblake I've been quite busy for the past couple of days. I'll review this in an hour and will let you know.

terranblake commented:
@liyasthomas no worries. No need to rush; just checking in

liyasthomas commented:
@terranblake 
* On `POST, PUT, PATCH` methods, while adding request body parameters, the default key value is appended as `[object MouseEvent]`.

![Screen Shot 2019-10-09 at 8 32 20 AM](https://user-images.githubusercontent.com/10395817/66448777-4ff1e380-ea70-11e9-89f5-548e900283e3.png)

terranblake commented:
@liyasthomas interesting. I'll take a look shortly

terranblake commented:
@liyasthomas this looks like a regression introduced by changing `addRequestBodyParam` to accept a default value when adding a new param. This should be fixed now 👍

liyasthomas commented:
@terranblake
* Use history feature seems to be broken.

terranblake commented:
@liyasthomas I didn't have any issues using history. Can you elaborate?

liyasthomas commented:
I'll test once again and let you know.

terranblake commented:
Merging latest changes from upstream master and will be re-testing this shortly @liyasthomas

terranblake commented:
Any specifics on what wasn't working with the use history feature?

terranblake commented:
@liyasthomas is it a little confusing to have 2 sets of `Parameter List` sections? There's the `bodyParams` section when `Raw Input` is disabled and the query params in the `Parameters` section

<img width="624" alt="Image 2019-10-14 at 7 29 34 PM" src="https://user-images.githubusercontent.com/8795767/66795431-6c02e400-eeb9-11e9-85b1-fb7b4181998b.png">
<img width="863" alt="Image 2019-10-14 at 7 30 04 PM" src="https://user-images.githubusercontent.com/8795767/66795435-6dcca780-eeb9-11e9-8c5e-483b2e606ae4.png">

liyasthomas commented:
@terranblake regarding body params and query params. Well, it's labels seems confusing. Will change to payload params and query params. Is that ok?

terranblake commented:
Yeah, or I can split them up in this PR. Just realized that those variables are shared at the moment, so we just need some separation between `queryParams` and `requestBody` 👍

liyasthomas commented:
You're good to go 👍

terranblake commented:
Awesome, thanks! 😝

liyasthomas commented on pages/index.vue:
```diff
@@ -791,15 +764,17 @@
-              url: this.url,
-              path: this.path
+              url,
+              path: url.pathname
```
On History section, `URL` contains `path`. Trim that down since `path` is already shown in separate field.

![Screen Shot 2019-10-15 at 10 49 24 AM](https://user-images.githubusercontent.com/10395817/66802446-10c50600-ef3a-11e9-9a74-7358ea0e5b6a.png)

terranblake commented:
@liyasthomas turns out I didn't notice `queryInputHandler` does everything to build the params correctly, so I was doing additional work that didn't need to be done, causing other issues.

All of this should be fixed and the history is looking correct now 🎈

liyasthomas commented:
* In query parameters section, adding new parms makes value `[object MouseEvent]` on key field.

* In same section, query parameters list doesn't reflect in `URL` field which was previously binded.

terranblake commented:
@liyasthomas I've updated the following things to resolve your above comments:
* `newRequestParam` now calls `addRequestParam` on `@click` without params; fixing the default `[object MouseEvent]` issue when adding new query parameters
* Bound the `Query Parameters` section to the URL field by updating the URL field whenever params are changed
* Introduced a test to validate that this continues to function as expected

NBTX commented:
@terranblake It looks like we're getting some merge conflicts. Can you rebase with `master`, please?

NBTX commented:
Also it seems when I delete a character from a query parameter box the URL box doesn't update. (It works the other way - URL box to query parameter box).

Just by eyeballing it, I'm going to suggest it's an issue related to `index.vue` line 813; you only update the URL box if `!this.url.includes(paramString)`, however deleting a character would return false for that.

NBTX commented on pages/index.vue:
```diff
@@ -830,21 +809,23 @@
-            pathParsed = url.parse(this.path);
-        return queryString = pathParsed.query ? pathParsed.query : '';
+      updateUrlParams(paramString = this.queryString) {
+        if (paramString && !this.url.includes(paramString)) {
```
Perhaps check the character that comes after the paramString, be it `&` or the end of the string - as then you know if that exact parameter string is contained in the URL - rather than relying on `includes`.

liyasthomas commented:
I afraid this became stale ☹️

NBTX commented:
It's pretty much complete, I can always just make the modification myself and submit a new PR.

liyasthomas commented:
@terranblake is currently working on evn vars. @NBTX could you make the modifications and double check the binding once again, because last time I checked I was having few bugs around import cURL and query params.

yubathom commented:
Hey @terranblake, I noticed that labels are not being tracked to the url. 
To fix this just add `'label'` to the _flats_ array at [line 971](https://github.com/terranblake/postwoman/blob/baf725331ecc9e173933a60e26c3eae25c7d04ab/pages/index.vue#L971) (please remove `'path'` from this array too). Some other checks:
 - I think you forgot a `debugger` in line [674](https://github.com/terranblake/postwoman/blob/baf725331ecc9e173933a60e26c3eae25c7d04ab/pages/index.vue#L674)
 - Since __path__ are not being used maybe it would be good to remove or review the usage of it at [424](https://github.com/terranblake/postwoman/blob/baf725331ecc9e173933a60e26c3eae25c7d04ab/pages/index.vue#424), [670](https://github.com/terranblake/postwoman/blob/baf725331ecc9e173933a60e26c3eae25c7d04ab/pages/index.vue#670), [675](https://github.com/terranblake/postwoman/blob/baf725331ecc9e173933a60e26c3eae25c7d04ab/pages/index.vue#675), [1008](https://github.com/terranblake/postwoman/blob/baf725331ecc9e173933a60e26c3eae25c7d04ab/pages/index.vue#1008) and [1077](https://github.com/terranblake/postwoman/blob/baf725331ecc9e173933a60e26c3eae25c7d04ab/pages/index.vue#1077).

@liyasthomas, I think would be good to write more tests. Could you describe the features that you are usually testing?

liyasthomas commented:
@yubathom code coverage is very low for us. Some of my manual tests includes:
- registering and reusing history entries
- cURL import (btw, there's a cURL export feature)
- query parameters bindings from `path` and parameters section
now we've Collections
- add, edit, delete Collections / folders / requests

I'm not sure what all could be included in tests 🙃

NBTX commented:
I think we definitely need a test for debugger lines and we should probably look into code coverage. Also I'm planning a major refactor of the index.vue page once we get some of the more major PRs landed as it's starting to grow quite messy.

liyasthomas commented:
@NBTX i think we should close this one. Terran seems to be busy af. And perhaps, if possible start a new one in future.

NBTX commented:
I will start this one again when I have time (most likely this weekend).
